### ISSUE TITLE
No performance gain with baichuan13B comparing with vllm

### ISSUE THREAD

**Issue by leiwen83** [ModelTC/lightllm#107]
Hi,

I am using baichuan13B to compare performance with vllm, and find in this case lightllm has no performance gain.
The same test vllm result is 240s, comparing with lightllm's result @252s.

Since baichuan13B has no fast tokenizer support, I am not sure whether it is the reason why performance bad? Or is there anything I could to tune to make it better?

**Comment by hiworldwzj** [comment]
@leiwen83  Fast tokenizer is needed to get high performance.

**Comment by hiworldwzj** [comment]
@leiwen83 what is your test params?

**Comment by leiwen83** [comment]
> @leiwen83 Fast tokenizer is needed to get high performance.

There are many new model don't support fast tokenizer, so for those model, and for those model finetuned without tokenizer.json, what can be done to automate the fast tokenizer conversion?
The reason for using fast tokenizer here I think is to accel the encode/decode process for "text->token id" and "token id->text", is that right?
So if it is true that lightllm is very sensitive towards the tokenizer encode/decode speed, is it better to sperate the encode/decode into another process, so maybe we could deploy encode/decode pod into some power cpu machine, and leave gpu machine only handle the pure token id without concerning for tokenizer performance?

**Comment by hiworldwzj** [comment]
@leiwen83 yes it is,  lightllm and vllm are sensitive towards the tokenizer encode/decode speed.

**Comment by hiworldwzj** [comment]
@leiwen83 What gpu card and setting did you use for testing?  When using a slow tokenizer, LightLLM should not be slower than VLLM either.

**Comment by leiwen83** [comment]
> @leiwen83 What gpu card and setting did you use for testing? When using a slow tokenizer, LightLLM should not be slower than VLLM either.

I am testing llama7B with 3090. I try increase --max_total_token_num param, seems performance is better, and could be par with vllm, also the same level.
I see the latest commit claim the performance is better with triton==2.1.0? I try triton master branch, but get several error, do you have some suggest triton commit to test with?

I am thinking maybe we could make another api interface to take raw encoded token as input/output without tokenizer involved, maybe in this way we could have a better performance gain comparing with vllm?

**Comment by hiworldwzj** [comment]
> > @leiwen83 What gpu card and setting did you use for testing? When using a slow tokenizer, LightLLM should not be slower than VLLM either.
> 
> I am testing llama7B with 3090. I try increase --max_total_token_num param, seems performance is better, and could be par with vllm, also the same level. I see the latest commit claim the performance is better with triton==2.1.0? I try triton master branch, but get several error, do you have some suggest triton commit to test with?
> 
> I am thinking maybe we could make another api interface to take raw encoded token as input/output without tokenizer involved, maybe in this way we could have a better performance gain comparing with vllm?
> 
> I check vllm code, it has raw token id input when use openai interface: https://github.com/vllm-project/vllm/blob/main/vllm/entrypoints/openai/api_server.py#L397
> 
> So it seems to me it is more make sense to test performance with openai interface?

@leiwen83  triton==2.1.0 has been tested in A800,  3090 has not been tested.  I can not give a suggest to you. triton==2.1.0 has not been officially released.  it may not be stable.   Maybe you can  adjust triton kernel setting(BLOCK or num_wrap in code) to faster the kernel.
